akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+ * level within the supplied timeout. Otherwise a [[Replicator.UpdateFailure]] subclass is
+ * sent back. Note that a [[Replicator.UpdateTimeout]] reply does not mean that the update completely failed
+ * or was rolled back. It may still have been replicated to some nodes, and will eventually
+ * be replicated to all nodes with the gossip protocol.
+ *
+ * You will always see your own writes. For example if you send two `Update` messages
+ * changing the value of the same `key`, the `modify` function of the second message will
+ * see the change that was performed by the first `Update` message.
+ *
+ * The `Update` message also supports a read consistency level with same meaning as described for
+ * `Get` below. If the given read consistency level is not `ReadLocal` it will first retrieve the
+ * data from other nodes and then apply the `modify` function with the latest data.
+ * If the read fails a [[Replicator.ReadFailure]] is replied to the sender of the `Update`. In the
+ * case of `ReadFailure` the update is aborted and no data has been changed.
+ * To support "read your own writes" all incoming commands for this key will be
+ * buffered until the read is completed and the function has been applied.
+ *
+ * In the `Update` message you can pass an optional request context, which the `Replicator`
+ * does not care about, but is included in the reply messages. This is a convenient
+ * way to pass contextual information (e.g. original sender) without having to use `ask`
+ * or local correlation data structures.
+ *
+ * == Get ==
+ *
+ * To retrieve the current value of a data you send [[Replicator.Get]] message to the
+ * `Replicator`. You supply a consistency level which has the following meaning:
+ * <ul>
+ * <li>`ReadLocal` the value will only be read from the local replica</li>
+ * <li>`ReadFrom(n)` the value will be read and merged from `n` replicas,
+ *     including the local replica</li>
+ * <li>`ReadQuorum` the value will be read and merged from a majority of replicas, i.e.
+ *     at least `N/2 + 1` replicas, where N is the number of nodes in the cluster
+ *     (or cluster role group)</li>
+ * <li>`ReadAll` the value will be read and merged from all nodes in the cluster
+ *     (or all nodes in the cluster role group)</li>
+ * </ul>
+ *
+ * As reply of the `Get` a [[Replicator.GetSuccess]] is sent to the sender of the
+ * `Get` if the value was successfully retrieved according to the supplied consistency
+ * level within the supplied timeout. Otherwise a [[Replicator.GetFailure]] is sent.
+ * If the key does not exist the reply will be [[Replicator.NotFound]].
+ *
+ * You will always read your own writes. For example if you send a `Update` message
+ * followed by a `Get` of the same `key` the `Get` will retrieve the change that was
+ * performed by the preceding `Update` message. However, the order of the reply messages are
+ * not defined, i.e. in the previous example you may receive the `GetSuccess` before
+ * the `UpdateSuccess`.
```

Comment:
All that I can say at this point is that I'm yet to stumble on a use-case for `ReadLocal`. That doesn't mean it is undesirable to have; it could also mean that I should be using it in places.